src/inference/src/ie_core.cpp

```diff
@@ -1159,69 +1190,78 @@ class CoreImpl : public ie::ICore, public std::enable_shared_from_this<ie::ICore
+                        desc.second.defaultConfig[conf.first] = conf.second;
+                    }
+                    configIsSet = true;
                 }
-                configIsSet = true;
             }
-        }
 
-        if (!configIsSet && !deviceName.empty()) {
-            IE_THROW() << "Device with \"" << deviceName << "\" name is not registered in the InferenceEngine";
-        }
-
-        // set config for already created plugins
-        for (auto& plugin : plugins) {
-            if (deviceName.empty() || clearDeviceName == plugin.first) {
-                allowNotImplemented([&]() {
-                    auto configCopy = config;
-                    if (DeviceSupportsCacheDir(plugin.second)) {
-                        auto cacheConfig = coreConfig.getCacheConfigForDevice(deviceName);
-                        if (cacheConfig._cacheManager) {
-                            configCopy[CONFIG_KEY(CACHE_DIR)] = cacheConfig._cacheDir;
-                        }
-                    } else if (configCopy.count(CONFIG_KEY(CACHE_DIR)) > 0) {
-                        // Remove "CACHE_DIR" from config if it is not supported by plugin
-                        configCopy.erase(CONFIG_KEY(CACHE_DIR));
-                    }
-                    // Add device specific value to support device_name.device_id cases
-                    std::vector<std::string> supportedConfigKeys =
-                        plugin.second.get_metric(METRIC_KEY(SUPPORTED_CONFIG_KEYS), {});
-                    auto config_iter = std::find(supportedConfigKeys.begin(),
-                                                 supportedConfigKeys.end(),
-                                                 CONFIG_KEY_INTERNAL(CONFIG_DEVICE_ID));
-                    const bool supportsConfigDeviceID = config_iter != supportedConfigKeys.end();
-                    const std::string deviceKey =
-                        supportsConfigDeviceID ? CONFIG_KEY_INTERNAL(CONFIG_DEVICE_ID) : CONFIG_KEY(DEVICE_ID);
+            if (!configIsSet && !deviceName.empty()) {
+                IE_THROW() << "Device with \"" << deviceName << "\" name is not registered in the InferenceEngine";
+            }
 
-                    if (!parser.getDeviceID().empty()) {
-                        configCopy[deviceKey] = parser.getDeviceID();
-                    }
-                    plugin.second.set_config(configCopy);
-                });
+            // set config for already created plugins
+            for (auto& plugin : plugins) {
+                if (deviceName.empty() || clearDeviceName == plugin.first) {
+                    created_plugins.emplace_back(
+                        std::pair<std::string, ov::InferencePlugin>{plugin.first, plugin.second});
+                }
             }
         }
+        for (auto& plugin : created_plugins) {
+            allowNotImplemented([&]() {
+                std::lock_guard<std::mutex> lock(get_mutex(plugin.first));
```

Comment:
why do we need to split the cycle into 2?
which function leads to recursive mutex lock?

Reply:
Yes, methods which work with device properties can achieve the global critical section.